support/builder/s3migrate.sh

```diff
@@ -0,0 +1,311 @@
+#!/bin/bash
+set -euo pipefail
+
+aggMsg() {
+    echo ""
+    echo "Aggregating locally stored artifacts."
+    echo "======================================"
+    echo ""
+
+}
+
+configMsg(){
+    echo ""
+    echo "Configuring environment for migration."
+    echo "======================================"
+    echo ""
+}
+
+keyMsg() {
+    echo ""
+    echo "Generating S3 Object Keys."
+    echo "======================================"
+    echo ""
+}
+
+uploadMsg() {
+    echo ""
+    echo "Uploading hartfile."
+    echo "======================================"
+}
+
+getHarts() {
+    # Traverses the depot data path and
+    # creates an array from each package
+    # artifact that exists in that path
+
+    TARGETDIR=${1}
+    for f in ${TARGETDIR}/*; do
+        if [[ -d "${f}" ]]; then
+            getHarts "${f}"
+        elif [[ "${f}" == *.hart ]]; then
+            echo "Found: $f"
+            artifacts+=("${f}")
+        else
+            echo "$f is not a hartfile, skipping."
+            echo ""
+        fi
+    done
+}
+
+uploadHarts() {
+    # Takes the artifact array generated via
+    # traversing the data path and uploads to s3
+    if [ ${#artifacts[@]} -eq 0 ]; then
+        echo ""
+        echo "No artifacts found in Artifact storage path!"
+        echo "Exiting!"
+        exit 1
+    fi
+        keyMsg
+    for hart in ${artifacts[@]}; do
+        uploadMsg
+        keyId=$(generateKey ${hart})
+        putArtifacts "${hart}" "${S3_BUCKET}" "${AWS_REGION}" "${keyId}"
+    done
+}
+
+installDeps() {
+    # We use aws-cli to handle the actual upload.
+    # but we also need jq to parse the pkg info
+
+    PDEPS=("core/aws-cli" "core/jq-static")
+
+    configMsg
+
+    for bin in "${!PDEPS[@]}"; do
+        hab pkg install -b ${PDEPS[$bin]}
+    done
+}
+
+generateKey() {
+    # Generates the s3 item key from the hartfile
+    FILEIN="${1}"
+    BASE="${FILEIN##*/}"
+
+    # Get the
+    ORIGIN="$(hab pkg info -j ${FILEIN} | jq -r .origin)"
+    PKGNAME="$(hab pkg info -j ${FILEIN} | jq -r .name)"
+    PKGVER="$(hab pkg info -j ${FILEIN} | jq -r .version)"
+    RELEASE="$(hab pkg info -j ${FILEIN} | jq -r .release)"
+    ARCH=("x86_64")
+
+    SYSTEMS=("linux" "windows")
+    for sys in "${SYSTEMS}"; do
+        if  [[ "${FILEIN}" = *"${sys}"* ]]; then
+            SYSTEM="${sys}"
+        fi
+    done
+
+    if [ -z $SYSTEM ]; then
+        return 1
+    else
+        echo "${ORIGIN}/${PKGNAME}/${PKGVER}/${RELEASE}/${ARCH}/${SYSTEM}/${BASE}"
+    fi
+}
+setBucket() {
+    # Configure the bucket name to use for the upload
+    echo ""
+    echo "==========================================================="
+    echo "Please enter a target bucket name and press [ENTER]:"
+    read bucket_name
```

Comment:
Instead of a target bucket name, it would be easier to just have them input the environment name.  Then we could just always have the bucket base name the same, and only change the environment as discussed.  That would be easier and likely less error prone than trying to remember the long bucket name.  We could also check for environments "live", "acceptance", or "default" and only allow those if we wanted.

Reply:
Unfortunately, this won't work. In AWS S3 bucket names have to be _globally_ unique. Which means you can't use a bucket by the same name as any other bucket. This tool quite unfortunately has to support a whole bunch of potential use cases :(

Reply:
Hmm, that's unfortunate.  This brings up one more concern - the migration script could mess up something already configured - eg, we mistakenly fat finger and type in the prod bucket when doing a dev test for example.  Perhaps there should be a check that migration doesn't start if the bucket exists AND there are already items in that bucket?

Reply:
Yeah I'll add a warning and confirmation on this!